pkg/driver/vol_statter.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package driver
+
+import (
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/volume/util/fs"
+	"time"
+)
+
+type volMetrics struct {
+	volPath string
+	timeStamp time.Time
+	volUsage []*csi.VolumeUsage
+}
+
+var (
+	volUsageCache = make(map[string]*volMetrics)
+	volStatterJobTracker = make (map[string]bool)
+)
+
+type VolStatter interface {
+	computeVolumeMetrics(volId, volPath string, refreshRate float64) (*volMetrics, error)
+	retrieveFromCache(volId string) (*volMetrics, bool)
+	removeFromCache(volId string)
+}
+
+type VolStatterImpl struct {
+}
+
+func NewVolStatter() VolStatter {
+	return &VolStatterImpl{}
+}
+
+func (v VolStatterImpl) computeVolumeMetrics(volId, volPath string, refreshRate float64) (*volMetrics, error){
+	if value, ok := v.retrieveFromCache(volId); ok {
+		if time.Since(value.timeStamp).Minutes() > refreshRate {
+			// Time to refresh volume stats
+			v.launchVolStatsRoutine(volId, volPath)
+		}
+		return value, nil
+	} else {
+		klog.V(4).Infof("Did not find volume metrics in cache for vol ID: %v , vol path: %v. Computing now!", volId, volPath)
+	}
+
+	v.launchVolStatsRoutine(volId, volPath)
+
+	// Return nil as kubelet might timeout waiting for volume stats
+	klog.Warningf("Volume metrics computation is underway for Vol ID: %v and metrics are not available yet.", volId)
+	return &volMetrics{
+		volPath: volPath,
+		timeStamp: time.Now(),
+		volUsage: []*csi.VolumeUsage{
+			{
+				Unit: csi.VolumeUsage_UNKNOWN,
+			},
+		},
+	}, nil
+}
+
+func (v VolStatterImpl) retrieveFromCache(volId string) (*volMetrics, bool) {
+	if value, ok := volUsageCache[volId]; ok {
+		return value, true
+	} else {
+		return nil, false
+	}
+}
+
+func (v VolStatterImpl) removeFromCache(volId string) {
+	delete(volUsageCache, volId)
+}
+
+func (v VolStatterImpl) launchVolStatsRoutine(volId, volPath string) {
+	if _, ok := volStatterJobTracker[volId]; ok {
+		klog.V(5).Infof("Volume stats computation job is underway for volume Id : %v. Awaiting results", volId)
+	} else {
+		volStatterJobTracker[volId] = true
+		go v.computeDiskUsage(volId, volPath)
+	}
+}
+
+func (v VolStatterImpl) computeDiskUsage(volId, volPath string) {
+	klog.V(5).Infof("Compute Volume Metrics invoked for Vol ID: %v", volId)
+
+	used, err := fs.DiskUsage(volPath)
```

Comment:
what happens if Unpublish is called on this volume in the middle of this routine? Would DiskUsage block the Unpublish unmount for the duration it's reading the volume?

Reply:
The logs from my testing indicate UnPublish will not succeed when DiskUsage is running. I see "device is busy" associated with Unpublish command in the logs. Do you think it's okay to wait it out?

Reply:
Darn, well I think it's okay to wait it out. We can leave it as a future improvement because we're relying on a library (fs.DiskUsage) and there's no way to cancel it.

In the worst-case, Unpublish could be delayed indefinitely if ` volume metrics period < computeDiskUsage execution time` and computeDiskUsage ends up running 100% of the time.

In a less-bad case, Unpublish could be delayed by `time it takes for DiskUsage to complete + 2 minutes`. The 2 minutes comes from the exponential back off for Unpublish retries. https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go#L33

Reply:
I'll add a TODO to figure out a way to cancel running DU process in the future